File: 0143-reorder-list/0143-reorder-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
    if(!head) return
    
    let slow = head
    let fast = head
    
    while(fast.next && fast.next.next){
        slow = slow.next
        fast = fast.next.next
    }
    
    let prev = null
    let curr = slow.next
    
    while (curr) {     
        let temp = curr.next
        curr.next = prev
        prev = curr
        curr = temp
    }
    slow.next = null

    let first = head
    let second = prev
    let temp = null
    while (second){
        temp = first.next
        first.next = second;
        first = temp;
        
        temp = second.next;
        second.next = first;
        second = temp;
    }
    console.log(head)
};

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# left-most right-most left-most right-most
#
#
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        second = slow.next
        slow.next = None
        prev = slow.next
        while second:
            tmp = second.next # stanard reverse the list 1. variable for next
            second.next = prev # 2. next pointer to prev
            prev = second # 3. prev set up to current for next iteratation
            second = tmp #4. move on via variable from step 1

        first, second = head, prev
        while second:
            temp1, temp2 = first.next, second.next
            first.next = second #1, 4
            second.next = temp1 #1,4->3
            first, second = temp1, temp2 # first = 2,3


        

        